### ISSUE TITLE
Implement `POST /opportunities` endpoint

### ISSUE THREAD

**Issue by slifty** [PhilanthropyDataCommons/service#33]
This will allow a third party to create a new opportunity in the system.  Note that authentication / authorization will be implemented separately from this issue.

**Comment by bickelj** [comment]
I expected TypeScript and AJV to narrow the type of `req.body` here such that it knows it has one property `title`, but no dice yet:
```TypeScript
  if (!isOpportunity(req.body)) {
    res.status(400)
      .contentType('application/json')
      .send(isOpportunity.errors ?? { message: 'Expected JSON body having title.' });
    return;
  }

  const opportunityPosted: Omit<Opportunity, 'createdAt' | 'id'> = { title: req.body.title };
```

I am probably missing something obvious, though.

**Comment by slifty** [comment]
Oh fascinating!  I hadn't thought about the "how does typescript handle type guards when run against attributes" question.  Yeah at a first glance I would have thought that to.

That being said I haven't crossed this path because destructuring objects that you plan on access ends up being the recommended approach (it's possible the linter will demand you do so, in fact)

Try this:

```
  const { body } = req;
  if (!isOpportunity(body)) {
    res.status(400)
      .contentType('application/json')
      .send(isOpportunity.errors ?? { message: 'Expected JSON body having title.' });
    return;
  }
  
  const opportunityPosted: Omit<Opportunity, 'createdAt' | 'id'> = { title: body.title };
```

Another fun fact -- you don't have to type on assignment (I don't know if the linter will tell you not to, but I suggest NOT doing that).  This means you can simply write:

```
  const opportunityPosted = { title: body.title };
```

**Comment by bickelj** [comment]
All of this is at compile/lint time, not run time, right. I suppose I can lean on `isOpportunity` rather than try to formally declare type expectations. In this case there is only one property of interest (`title`) but in other cases (other entities) there will be several. Also I expect `Opportunity` (or others) to evolve, so the goal was to have the type checker flag those kinds of code changes needed when a change occurs. Perhaps it can be caught elsewhere, too, though, or in a better way.

With `const { body } = req;` and later `  const opportunityPosted: Omit<Opportunity, 'createdAt' | 'id'> = { title: body.title };`:
> Unsafe array destructuring of a tuple element with an `any` value  @typescript-eslint/no-unsafe-assignment

I think I would rather say something like `const opportunityPosted: Omit<Opportunity, 'createdAt' | 'id'> = body;` but perhaps I am asking too much of TypeScript and AJV.

As for `opportunityPosted` that is in contrast to a later variable declared, `opportunityInserted` which is a full (read-only) `Opportunity` rather than a partial (write) `Omit<Opportunity, 'createdAt' | 'id'>`.

**Comment by bickelj** [comment]
If I relax the schema a bit to look like this:
```TypeScript

export interface Opportunity {
  id?: number;
  title: string;
  createdAt?: Date;
}

export const opportunitySchema: JSONSchemaType<Opportunity> = {
  type: 'object',
  properties: {
    id: {
      type: 'integer',
      nullable: true,
    },
    title: {
      type: 'string',
    },
    createdAt: {
      type: 'object',
      required: [],
      instanceof: 'Date',
      nullable: true,
    },
  },
  required: [
    'title',
  ],
};
```
And then have a POST handler like this:
```TypeScript

const postOpportunity = (req: Request, res: Response): void => {
  logger.debug(req);

  if (!isOpportunity(req.body)) {
    res.status(400)
      .contentType('application/json')
      .send(isOpportunity.errors ?? { message: 'Expected JSON body having title.' });
    return;
  }

  db.sql('opportunities.insertOne', req.body)
    .then((opportunityInserted: Result<Opportunity>) => {
      logger.debug(opportunityInserted);
      if (opportunityInserted.row_count === 0) {
        res.status(500)
          .contentType('application/json')
          .send({ message: 'Unable to find newly created opportunity.' });
        return;
      }

      const { rows } = opportunityInserted;

      if (isOpportunity(rows[0])) {
        res.status(201)
          .location(`${req.url}/${rows[0].id}`)
          .contentType('application/json')
          .send(rows[0]);
      } else {
        throw new ValidationError(
          'The database responded with an unexpected format.',
          isOpportunity.errors ?? [],
        );
      }
    })
    .catch((error: unknown) => {
      logger.error(error);
      res.status(500)
        .contentType('application/json')
        .send(error);
    });
};
```

Then it all works wonderfully. Except now that `id` is optional, this line has a linter warning, and rightly so, ```.location(`${req.url}/${rows[0].id}`)```, because id could be undefined. However in this context we know the database will come back with a full opportunity.

In another channel we talked about seeing if AJV has something akin to utility types for the schemas. I think that would be most satisfying. There might be other ways around it too.

**Comment by bickelj** [comment]
A type guard is enough.
```TypeScript
      if (isOpportunity(rows[0])) {
        res.status(201)
          .location(`${req.url}/${rows[0].id}`)
```
becomes
```TypeScript
      if (isOpportunity(rows[0]) && typeof rows[0].id === 'number') {
        res.status(201)
          .location(`${req.url}/${rows[0].id}`)
```
Good enough for now!

**Comment by bickelj** [comment]
Resolved with merge of #70
